sw/otbn/crypto/p384_modinv.s

```diff
@@ -0,0 +1,87 @@
+/* Copyright lowRISC contributors. */
+/* Licensed under the Apache License, Version 2.0, see LICENSE for details. */
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ *   P-384 specific routines for ECC modular inverse computation.
+ */
+
+ .section .text
+
+/**
+ * Variable-time modular multiplicative inverse computation
+ *
+ * returns x_inv = x^-1 mod m
+ *
+ * This routine computes the modular multiplicative inverse for any x < m in
+ * the finite field GF(m) where m is prime.
+ *
+ * For inverse computation, Fermat's little theorem is used, i.e.
+ * we compute x^-1 = x^(m-2) mod m.
+ * For exponentiation we use a standard, variable-time (!) square and multiply
+ * algorithm.
+ *
+ * This routine is mainly intended to be used for inversion of scalars in
+ * context of the P-384 curve. In theory, it can be used with any 384-bit
+ * modulus m with a corresponding 385-bit Barrett constant u,
+ * where u[383:192] = 0.
+ *
+ * Note: When used for P-384 scalar inversion, the routine will need 672 calls
+ * to the multiplication routine. By using an adder chain this could be reduced
+ * to ~433 multiplications, however, at the cost of a significant codes size
+ * increase.
+ *
+ * Note: This routine runs in variable-time w.r.t. the modulus. It should only
+ * be used with a non-secret modulus.
+ *
+ * @param[in]  [w13, w12]: m, 384 bit modulus
+ * @param[in]  w14: k, Solinas constant (2^384 - m) (max. length 191 bits).
+ * @param[in]  [w30, w29]: x, 384 bit operand
+ * @param[in]  w31, all-zero
+ * @param[out] [w17, w16]: x_inv, modular multiplicative inverse
+ *
+ * Flags: Flags have no meaning beyond the scope of this subroutine.
+ *
+ * clobbered registers: x2, w2, w3, w10, w11, w16 to w24
+ * clobbered flag groups: FG0
+ */
+ .globl mod_inv_n_p384
+mod_inv_n_p384:
+
+  /* subtract 2 from modulus for Fermat's little theorem
+     [w13,w12] <= m - 2 = [w11,w10]-2 (left aligned) */
+  bn.subi   w2, w12, 2
+  bn.subb   w3, w13, w31
+  bn.rshi   w3, w3, w2 >> 128
```

Comment:
If `w31` is zero, isn't this the same as `bn.rshi w3, w13, w2 >> 128` ?

I'm asking because I might be wrong! And I'd like to understand what I've got confused about :-)

Reply:
Yes this would be the same, but only if no carry flag has been set in the previous subtraction `bn.subi w2, w12, 2` in the line above (line 52). This in the end depends on modulus function parameter `m`. In case we assume that the only two possible values for `m` are `n` and `p`, a carry flag would never get set, because the lower 256 bits of `n`/`p` are greater than `2`. In this case we could replace the two lines by `bn.rshi w3, w13, w2 >> 128`.
Maybe the original author of this function wanted it to be more generic, in case of other modulus than `n` or `p`. I didn't pay attention to it so far, as I just moved the code to a different file ;) .

Reply:
Ahh! I hadn't thought about the carry flag! Thank you very much for the explanation.